src/v/storage/disk_log_impl.cc

```diff
@@ -557,43 +557,71 @@ model::term_id disk_log_impl::term() const {
+    if (ret.start_offset > model::offset(0)) {
+        ret.dirty_offset = ret.start_offset - model::offset(1);
+        ret.committed_offset = ret.dirty_offset;
+    }
+    return ret;
+}
 
 offset_stats disk_log_impl::offsets() const {
     if (_segs.empty()) {
-        offset_stats ret;
-        ret.start_offset = _start_offset;
-        if (ret.start_offset > model::offset(0)) {
-            ret.dirty_offset = ret.start_offset - model::offset(1);
-            ret.committed_offset = ret.dirty_offset;
-        }
-        return ret;
+        return empty_offsets(_start_offset);
     }
-    // NOTE: we have to do this because ss::circular_buffer<> does not provide
-    // with reverse iterators, so we manually find the iterator
+
     segment_set::type end;
     segment_set::type term_start;
+    segment_set::type last_committed;
+    /**
+     * We have to iterate backward over segments set as we have to find last
+     * non empty segment, last segment with initialized committed offset and
+     * first offset of current term
+     * NOTE: we have to do this because ss::circular_buffer<> does not
+     *       provide with reverse iterators, so we manually find the iterator
+     */
     for (int i = (int)_segs.size() - 1; i >= 0; --i) {
         auto& seg = _segs[i];
-        if (!seg->empty()) {
-            if (!end) {
-                end = seg;
-            }
-            // find term start offset
-            if (seg->offsets().term < end->offsets().term) {
+        // skip over empty segment
+        if (seg->empty()) {
+            continue;
+        }
+
+        // last not empty segment
+        if (!end) {
+            end = seg;
+        }
+
+        const auto& segment_offsets = seg->offsets();
+        // replace last_committed until we find segment with positive committed
+        // offset
+        if (
+          !last_committed
+          || last_committed->offsets().committed_offset < model::offset(0)) {
+            last_committed = seg;
+        }
+
+        // we prevent replacing term_start pointer if term is smaller than end
+        // segment term, this way term_start points to the first segment with
+        // current term
+        if (term_start && segment_offsets.term < end->offsets().term) {
+            // found both, stop iterating
+            if (
+              last_committed->offsets().committed_offset >= model::offset(0)) {
                 break;
             }
-            term_start = seg;
+            // still missing segment containing last committed offset
+            continue;
         }
+        term_start = seg;
     }
+
     if (!end) {
-        offset_stats ret;
-        ret.start_offset = _start_offset;
-        if (ret.start_offset > model::offset(0)) {
-            ret.dirty_offset = ret.start_offset - model::offset(1);
-            ret.committed_offset = ret.dirty_offset;
-        }
-        return ret;
+        return empty_offsets(_start_offset);
     }
+
```

Comment:
Is there a reason not to implement it like this:
```suggestion
    /**
     * We have to iterate backward over segments set as we have to find last
     * non empty segment, last segment with initialized committed offset and
     * first offset of current term
     */
    const auto rbegin = std::make_reverse_iterator(_segs.cbegin());
    const auto rend = std::find_if(
      std::make_reverse_iterator(_segs.cend()), rbegin, [](const auto& seg) {
          return !seg->empty();
      });

    // If all segments are empty, return empty offsets.
    if (rend == rbegin) {
        return empty_offsets(_start_offset);
    }

    auto it_last_committed = std::find_if(rend, rbegin, [](const auto& seg) {
        return seg->offsets().committed_offset >= model::offset(0);
    });

    // If nothing was committed, use the first segment
    const auto& last_committed = it_last_committed != rbegin
                                   ? *it_last_committed
                                   : _segs.front();

    auto it_term_start = std::lower_bound(
      _segs.begin(),
      _segs.end(),
      (*rend)->offsets().term,
      [](const auto& seg, model::term_id term) {
          return seg->offsets().term < term;
      });

    // If term_start not found, use the first segment
    const auto& term_start = it_term_start != _segs.end() ? *it_term_start
                                                          : _segs.front();
```

Reply:
This requires iterating over segments set few times

Reply:
Yes, but my thoughts are:
* The code is much easier to read
* offsets are at the end of a pointer, getting that into cache probably dominates, so binary search is probably faster than linear.
* I have no a priori knowledge about whether last_committed and start_term are correlated in space

Reply:
agreed, I will change this

Reply:
@BenPope  I would rename `rbegin` with `rend`. now it seems confusing

Reply:
actually it is less confusing with proposed naming

Reply:
last thing that bothers me is the `std::lower_bound` it iterates from the begging of segment set

Reply:
@mmaslankaprv if we make the segment_set random-access, we get that in `log N`

Reply:
> last thing that bothers me is the `std::lower_bound` it iterates from the begging of segment set

It starts in the middle, but we could write one that takes an initial pivot and/or does an interpolation search. What would you provide as a hint?

Reply:
> @mmaslankaprv if we make the segment_set random-access, we get that in `log N`

Iterator on `_segs` are random access (it's backed by a circular_buffer).